File: backend/src/models/DailySnapshot.ts
import * as admin from 'firebase-admin';
import { db } from '../global/firebase';

// ── 型別定義 ────────────────────────────────────────────────────────────────

export interface DailySnapshotInput {
  date: string;             // YYYY-MM-DD（台灣時間）
  totalInvested: number;
  stockValue: number;
  cashBalance: number;
  forexValue: number;
  unrealizedProfit: number;
  realizedProfit: number;
  totalReturn: number;
  returnRate: number;       // e.g. 0.1371（小數，非百分比）
  note?: string;
}

export interface DailySnapshotDoc extends Required<DailySnapshotInput> {
  recordedAt: Date;
}

export interface SnapshotPatch {
  cashBalance?: number;
  note?: string;
}

const COL = 'daily_snapshots';

// ── Model ───────────────────────────────────────────────────────────────────

export class DailySnapshot {
  /** 依日期範圍查詢，降序排列 */
  static async findByRange(from: string, to: string): Promise<DailySnapshotDoc[]> {
    const snap = await db
      .collection(COL)
      .where('date', '>=', from)
      .where('date', '<=', to)
      .orderBy('date', 'desc')
      .get();
    return snap.docs.map(d => deserialize(d));
  }

  /** 查詢單日快照 */
  static async findByDate(date: string): Promise<DailySnapshotDoc | null> {
    const doc = await db.collection(COL).doc(date).get();
    return doc.exists ? deserialize(doc) : null;
  }

  /** 查詢最新一筆快照（用於取得上次 cash_balance） */
  static async findLatest(): Promise<DailySnapshotDoc | null> {
    const snap = await db
      .collection(COL)
      .orderBy('date', 'desc')
      .limit(1)
      .get();
    if (snap.empty) return null;
    return deserialize(snap.docs[0]);
  }

  /** 寫入快照（merge 冪等設計，同日重複觸發只更新不重複寫） */
  static async record(input: DailySnapshotInput): Promise<DailySnapshotDoc> {
    const ref = db.collection(COL).doc(input.date);
    await ref.set(
      {
        date:              input.date,
        total_invested:    input.totalInvested,
        stock_value:       input.stockValue,
        cash_balance:      input.cashBalance,
        forex_value:       input.forexValue,
        unrealized_profit: input.unrealizedProfit,
        realized_profit:   input.realizedProfit,
        total_return:      input.totalReturn,
        return_rate:       input.returnRate,
        note:              input.note ?? '',
        recorded_at:       admin.firestore.FieldValue.serverTimestamp(),
      },
      { merge: true }
    );
    const updated = await ref.get();
    return deserialize(updated);
  }

  /** 修正活存 / 備註（PUT 用） */
  static async update(date: string, patch: SnapshotPatch): Promise<DailySnapshotDoc | null> {
    const ref      = db.collection(COL).doc(date);
    const existing = await ref.get();
    if (!existing.exists) return null;

    const data: Record<string, unknown> = {};
    if (patch.cashBalance !== undefined) data.cash_balance = patch.cashBalance;
    if (patch.note        !== undefined) data.note         = patch.note;

    if (Object.keys(data).length === 0) return deserialize(existing);
    await ref.update(data);
    const updated = await ref.get();
    return deserialize(updated);
  }
}

// ── 反序列化 ─────────────────────────────────────────────────────────────────

function deserialize(doc: admin.firestore.DocumentSnapshot): DailySnapshotDoc {
  const d = doc.data()!;
  return {
    date:             d.date,
    totalInvested:    d.total_invested,
    stockValue:       d.stock_value,
    cashBalance:      d.cash_balance,
    forexValue:       d.forex_value,
    unrealizedProfit: d.unrealized_profit,
    realizedProfit:   d.realized_profit,
    totalReturn:      d.total_return,
    returnRate:       d.return_rate,
    note:             d.note ?? '',
    recordedAt:       d.recorded_at instanceof admin.firestore.Timestamp
      ? d.recorded_at.toDate()
      : new Date(),
  };
}

File: Back-End/shioaji-service/models/daily_snapshot.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "daily_snapshots"


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "date": d.get("date", ""),
        "totalInvested": d.get("total_invested", 0),
        "stockValue": d.get("stock_value", 0),
        "cashBalance": d.get("cash_balance", 0),
        "forexValue": d.get("forex_value", 0),
        "unrealizedProfit": d.get("unrealized_profit", 0),
        "realizedProfit": d.get("realized_profit", 0),
        "totalReturn": d.get("total_return", 0),
        "returnRate": d.get("return_rate", 0),
        "note": d.get("note", ""),
        "recordedAt": _ts_to_iso(d.get("recorded_at")),
    }


async def find_all(year: Optional[int] = None) -> list[dict]:
    def _run():
        from_date = f"{year}-01-01" if year else "2000-01-01"
        to_date   = f"{year}-12-31" if year else "9999-12-31"
        snap = (
            get_db().collection(COL)
            .where("date", ">=", from_date)
            .where("date", "<=", to_date)
            .order_by("date", direction="DESCENDING")
            .get()
        )
        return [_from_doc(d) for d in snap]
    return await asyncio.to_thread(_run)


async def find_by_range(from_date: str, to_date: str) -> list[dict]:
    def _run():
        snap = (
            get_db().collection(COL)
            .where("date", ">=", from_date)
            .where("date", "<=", to_date)
            .order_by("date", direction="DESCENDING")
            .get()
        )
        return [_from_doc(d) for d in snap]
    return await asyncio.to_thread(_run)


async def find_by_date(date: str) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(date).get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def find_latest() -> Optional[dict]:
    def _run():
        snap = (
            get_db().collection(COL)
            .order_by("date", direction="DESCENDING")
            .limit(1)
            .get()
        )
        docs = list(snap)
        return _from_doc(docs[0]) if docs else None
    return await asyncio.to_thread(_run)


async def record(data: dict) -> dict:
    def _run():
        ref = get_db().collection(COL).document(data["date"])
        ref.set({
            "date": data["date"],
            "total_invested": data["totalInvested"],
            "stock_value": data["stockValue"],
            "cash_balance": data["cashBalance"],
            "forex_value": data["forexValue"],
            "unrealized_profit": data["unrealizedProfit"],
            "realized_profit": data["realizedProfit"],
            "total_return": data["totalReturn"],
            "return_rate": data["returnRate"],
            "note": data.get("note", ""),
            "recorded_at": SERVER_TIMESTAMP,
        }, merge=True)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def update(date: str, patch: dict) -> Optional[dict]:
    def _run():
        ref = get_db().collection(COL).document(date)
        if not ref.get().exists:
            return None
        data: dict = {}
        if "cashBalance" in patch: data["cash_balance"] = patch["cashBalance"]
        if "note" in patch:        data["note"]         = patch["note"]
        if data:
            ref.update(data)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)
